### PR TITLE
[Syntax] Add CRLF support to libSyntax

### PR DESCRIPTION
**Pull request by omochi**
This PR add CRLF support to libSyntax.
CRLF is added as new Trivia which is `TriviaKind::CarriageReturnLineFeed`.

To write simply parse logic in Lexer,
I added `trySquash` method to `TriviaPiece`,
`appendOrSquash` method to `Trivia`.

This is process that compose TriviaKind which keeps length by Count
such like Newline or CarriageReturnLineFeed to one TriviaKind and increase Count.

Then, I changed type of `LeadingTrivia` and `TrailingTrivia` in Lexer from
`TriviaList` to `Trivia`.

### CODE REVIEW

File: lib/Syntax/Trivia.cpp [hunk 1]
```diff
@@ -101,6 +104,7 @@ void TriviaPiece::accumulateAbsolutePosition(AbsolutePosition &Pos) const {
     break;
   case TriviaKind::Newline:
   case TriviaKind::CarriageReturn:
+  case TriviaKind::CarriageReturnLineFeed:
     Pos.addNewlines(Count);
```
Comment (rintaro) [inline lib/Syntax/Trivia.cpp]: `addNewlines(Count)` only increments `Pos.Offset` by `Count`.
We also should fix `addText(Str)` to take `\r` and `\r\n` into account
see: https://github.com/apple/swift/blob/3f0e7029c5d09ffa2a9daf54e9715bcc068b420e/include/swift/Syntax/RawSyntax.h#L139-L162

File: lib/Parse/Lexer.cpp [hunk 2]
```diff
@@ -2302,42 +2302,34 @@ void Lexer::lexTrivia(syntax::TriviaList &Pieces, bool IsForTrailingTrivia) {
...
-    // Consume consective same characters.
-    while (*CurPtr == Char)
-      ++CurPtr;
-
-    auto Length = CurPtr - TriviaStart;
-    switch (Char) {
-    case ' ':
-      Pieces.push_back(TriviaPiece::spaces(Length));
-      break;
+    switch (CurPtr[-1]) {
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: We can assume that most people only use `\n`.  So, I think, probably
```cpp
case '\n'
  if (IsForTrailingTrivia)
    break;
  NextToken.setAtStartOfLine(true);
  Pieces.appendOrSquash(TriviaPiece::newlines(1));
case '\r':
  if (IsForTrailingTrivia)
    break;
  NextToken.setAtStartOfLine(true);
  if (*CurPtr == '\n') {
    Pieces.appendOrSquash(TriviaPiece::carriageReturnLineFeeds(1));
```
is slightly more efficient.

File: lib/Parse/Lexer.cpp [hunk 3]
```diff
@@ -2302,42 +2302,34 @@ void Lexer::lexTrivia(syntax::TriviaList &Pieces, bool IsForTrailingTrivia) {
...
-      Pieces.push_back(TriviaPiece::tabs(Length));
-      break;
-    case '\v':
-      Pieces.push_back(TriviaPiece::verticalTabs(Length));
-      break;
-    case '\f':
-      Pieces.push_back(TriviaPiece::formfeeds(Length));
+      if (CurPtr[0] == '\n') {
+        Pieces.appendOrSquash(TriviaPiece::carriageReturnLineFeeds(1));
+        CurPtr++;
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: Nitpick: We usually use `++CurPtr` for simple increment.

File: lib/Parse/Lexer.cpp (hunk 2 above)
Comment (omochi) [reply lib/Parse/Lexer.cpp]: I see. I change so.

File: lib/Syntax/Trivia.cpp (hunk 1 above)
Comment (omochi) [reply lib/Syntax/Trivia.cpp]: I will read around there.

File: lib/Parse/Lexer.cpp (hunk 3 above)
Comment (omochi) [reply lib/Parse/Lexer.cpp]: ok. I will fix this.

**Comment by omochi** [comment]
Thanks to review.
I will make new commit.

File: lib/Parse/Lexer.cpp [hunk 4]
```diff
@@ -2298,46 +2298,34 @@ void Lexer::lexTrivia(syntax::TriviaList &Pieces, bool IsForTrailingTrivia) {
   // TODO: Handle invalid UTF8 sequence which is skipped in lexImpl().
   switch (*CurPtr++) {
   case '\n':
+    if (IsForTrailingTrivia)
+      break;
+    NextToken.setAtStartOfLine(true);
+    Pieces.appendOrSquash(TriviaPiece::newlines(1));
+    break;
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: `goto Restart;`?

Comment (omochi) [reply lib/Parse/Lexer.cpp]: I mistook.
I will fix it and add regression test.

Comment (omochi) [reply lib/Parse/Lexer.cpp]: Ah I figured out why this passed existing tests.
After break, lexTrivia return with CurPtr reset.
Then in lexImpl, '\n' at CurPtr[0] is skipped as whitespace.
And when reenter lexTrivia, all state is valid for continue. 

I will fix it.

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by omochi** [comment]
I have some questions.

1. Why did bot test not 70986a6 which is pushed before when you order to bot but f86e1c8.
2. Test failure reason of f86e1c8 seems to below and too small.

```
[swift-PR-osx] $ /bin/bash -e /var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/jenkins5647438479677286650.sh
Wrong sha f86e1c82015680d50626cd87af080e7aee1f0e39
```

I often see "Wrong sha XXX" with git commit which was not pushed but this commit obviously exists on server.

Did I wrong something?

**Comment by rintaro** [comment]
"Wrong sha XXX" failure is not your fault, but https://github.com/jenkinsci/ghprb-plugin/issues/414
The bot automatically reruns the build.

**Comment by rintaro** [comment]
Thank you @omochi !

**Comment by omochi** [comment]
Okey about "Wrong sha XXX".
Thanks to review and merge!

**Pull request merged**